### PR TITLE
Cleanup genesis loader

### PR DESCRIPTION
**Pull request by diega**
Now all the repository handling is made only in the `org.ethereum.core.genesis.BlockChainLoader#loadBlockchain` method. This method will be adjusted in the `unitrie` branch for removing `ContractDetails`

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 12 issues
* ![CRITICAL][CRITICAL] 4 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 1 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [BlockChainLoader.java#L27](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L27): Unused import - co.rsk.db.ContractDetailsImpl. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [BlockChainLoader.java#L28](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L28): Unused import - co.rsk.trie.TrieImpl. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [BlockChainLoader.java#L29](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L29): Unused import - co.rsk.trie.TrieStoreImpl. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![CRITICAL][CRITICAL] [BlockChainLoader.java#L33](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L33): Unused import - org.ethereum.datasource.HashMapDB. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![MAJOR][MAJOR] [BlockChainLoader.java#L145](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L145): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [GenesisLoader.java#L62](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L62): This method new java.util.Scanner(InputStream, String) should use a StandardCharsets.XXX.name() to specify an encoding [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACSI_CHAR_SET_ISSUES_USE_STANDARD_CHARSET_NAME)
1. ![MAJOR][MAJOR] [GenesisLoader.java#L129](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L129): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [GenesisLoader.java#L130](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L130): Method org.ethereum.core.genesis.GenesisLoader.mapFromJson(BigInteger, GenesisJson, boolean) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MINOR][MINOR] [BlockChainLoader.java#L71](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L71): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MINOR][MINOR] [BlockChainLoader.java#L140](https://github.com/rsksmart/rskj/blob/ad2c199487f30dc0d6f36144730831633fe3dfb0/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L140): Remove this use of "getContractDetails"; it is deprecated. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 6 issues

1. ![MAJOR][MAJOR] [BlockChainLoader.java#L141](https://github.com/rsksmart/rskj/blob/ee1d94c3fa2848153b3ba1f84f90106bd8b37d1e/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L141): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [GenesisLoader.java#L130](https://github.com/rsksmart/rskj/blob/ee1d94c3fa2848153b3ba1f84f90106bd8b37d1e/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L130): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MINOR][MINOR] [BlockChainLoader.java#L67](https://github.com/rsksmart/rskj/blob/ee1d94c3fa2848153b3ba1f84f90106bd8b37d1e/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L67): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MINOR][MINOR] [BlockChainLoader.java#L136](https://github.com/rsksmart/rskj/blob/ee1d94c3fa2848153b3ba1f84f90106bd8b37d1e/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L136): Remove this use of "getContractDetails"; it is deprecated. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [GenesisLoader.java#L60](https://github.com/rsksmart/rskj/blob/ee1d94c3fa2848153b3ba1f84f90106bd8b37d1e/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L60): Remove this unused method parameter "config". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1172)
1. ![INFO][INFO] [GenesisLoader.java#L63](https://github.com/rsksmart/rskj/blob/ee1d94c3fa2848153b3ba1f84f90106bd8b37d1e/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L63): Static or private method org.ethereum.core.genesis.GenesisLoader.loadGenesis(RskSystemProperties, BigInteger, InputStream, boolean) has unused parameters [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AUP_UNUSED_PARAMETER)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 7 issues

1. ![MAJOR][MAJOR] [RskFactory.java#L326](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L326): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [RskFactory.java#L327](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L327): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [RskFactory.java#L333](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L333): java/io/FileOutputStream.<init>(Ljava/lang/String;)V writes to a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_OUT)
1. ![MAJOR][MAJOR] [BlockChainLoader.java#L141](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L141): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [GenesisLoader.java#L120](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L120): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MINOR][MINOR] [BlockChainLoader.java#L67](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L67): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MINOR][MINOR] [BlockChainLoader.java#L136](https://github.com/rsksmart/rskj/blob/67e110f1f5ef224a9e75f6bb48bd7494419393b5/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L136): Remove this use of "getContractDetails"; it is deprecated. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java [hunk 1]
```diff
@@ -121,20 +124,31 @@ public BlockChainImpl loadBlockchain() {
...
-                repository.createAccount(addr);
-                InitialAddressState initialAddressState = genesis.getPremine().get(addr);
-                repository.addBalance(addr, initialAddressState.getAccountState().getBalance());
-                AccountState accountState = repository.getAccountState(addr);
-                accountState.setNonce(initialAddressState.getAccountState().getNonce());
-
-                if (initialAddressState.getContractDetails()!=null) {
-                    repository.updateContractDetails(addr, initialAddressState.getContractDetails());
-                    accountState.setStateRoot(initialAddressState.getAccountState().getStateRoot());
-                    accountState.setCodeHash(initialAddressState.getAccountState().getCodeHash());
```
Comment (lsebrie) [inline rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java]: We should keep this line to set codeHash and keep consistency

File: rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java [hunk 2]
```diff
@@ -121,20 +124,31 @@ public BlockChainImpl loadBlockchain() {
...
+
+            // second we create contracts whom only have code modifying the preexisting ContractDetails instance
+            for (Map.Entry<RskAddress, byte[]> codeEntry : genesis.getCodes().entrySet()) {
+                RskAddress contractAddress = codeEntry.getKey();
+                ContractDetails contractDetails = repository.getContractDetails(contractAddress);
+                contractDetails.setCode(codeEntry.getValue());
+                Map<DataWord, byte[]> contractStorage = genesis.getStorages().get(contractAddress);
+                for (Map.Entry<DataWord, byte[]> storageEntry : contractStorage.entrySet()) {
+                    byte[] storageValue = storageEntry.getValue();
+                    if (storageValue == null) {
```
Comment (lsebrie) [inline rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java]: I wouldn't do kind of check, if the storage values are wrong is better to know when they are inserted. I would remove the if condition

File: rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java [hunk 3]
```diff
@@ -77,50 +68,81 @@ public static Genesis loadGenesis(RskSystemProperties config, BigInteger initial
...
-    private static byte[] generateRootHash(Map<RskAddress, InitialAddressState> premine){
-        Trie state = new TrieImpl(null, true);
-
-        for (RskAddress addr : premine.keySet()) {
-            state = state.put(addr.getBytes(), premine.get(addr).getAccountState().getEncoded());
+    private static byte[] calculateStateRoot(Map<DataWord, byte[]> contractStorage) {
+        Trie trie = new TrieImpl(true);
+        for (Map.Entry<DataWord, byte[]> storageEntry : contractStorage.entrySet()) {
+            byte[] storageValue = storageEntry.getValue();
+            if (storageValue == null) {
```
Comment (lsebrie) [inline rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java]: This is the same case as the one above and storage

File: rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java (hunk 1 above)
Comment (diega) [reply rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java]: All this behaviour has been moved into the `GenesisLoader#mapFromJson` logic, the actual `AccountState` is created in the _loader_ with the proper _codeHash_ in place

**Comment by rskops** [comment]
SonarQube analysis reported 6 issues

1. ![MAJOR][MAJOR] [RskFactory.java#L326](https://github.com/rsksmart/rskj/blob/c3fcdbe76665312fe0cdb57bf02a1e164b23efd9/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L326): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [RskFactory.java#L327](https://github.com/rsksmart/rskj/blob/c3fcdbe76665312fe0cdb57bf02a1e164b23efd9/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L327): java/nio/file/Paths.get(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path; reads a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_IN)
1. ![MAJOR][MAJOR] [RskFactory.java#L333](https://github.com/rsksmart/rskj/blob/c3fcdbe76665312fe0cdb57bf02a1e164b23efd9/rskj-core/src/main/java/co/rsk/core/RskFactory.java#L333): java/io/FileOutputStream.<init>(Ljava/lang/String;)V writes to a file whose location might be specified by user input [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findsecbugs%3APATH_TRAVERSAL_OUT)
1. ![MAJOR][MAJOR] [GenesisLoader.java#L120](https://github.com/rsksmart/rskj/blob/c3fcdbe76665312fe0cdb57bf02a1e164b23efd9/rskj-core/src/main/java/org/ethereum/core/genesis/GenesisLoader.java#L120): Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS134)
1. ![MINOR][MINOR] [BlockChainLoader.java#L67](https://github.com/rsksmart/rskj/blob/c3fcdbe76665312fe0cdb57bf02a1e164b23efd9/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L67): Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00107)
1. ![MINOR][MINOR] [BlockChainLoader.java#L136](https://github.com/rsksmart/rskj/blob/c3fcdbe76665312fe0cdb57bf02a1e164b23efd9/rskj-core/src/main/java/org/ethereum/core/genesis/BlockChainLoader.java#L136): Remove this use of "getContractDetails"; it is deprecated. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**